stdlib/Distributed/src/remotecall.jl

```diff
@@ -548,11 +555,40 @@ Further calls to `fetch` on the same reference return the cached value. If the r
 is an exception, throws a [`RemoteException`](@ref) which captures the remote exception and backtrace.
 """
 function fetch(r::Future)
-    r.v !== nothing && return something(r.v)
-    v = call_on_owner(fetch_ref, r)
-    r.v = Some(v)
+    v_cache = @atomic r.v
+    v_cache !== nothing && return something(v_cache)
+
+    if r.where == myid()
+        v_cache, rv = lock(r.lock) do
+            v_cache = @atomic :monotonic r.v
+            v_cache, v_cache === nothing ? lookup_ref(remoteref_id(r)) : nothing
+        end
+
+        if v_cache !== nothing
+            return something(v_cache)
+        else
+            v_local = fetch(rv.c)
+        end
+    else
+        v_local = call_on_owner(fetch_ref, r)
+    end
+
+    v_cache = @atomic r.v
+
+    if v_cache === nothing # call_on_owner case
+        v_old, status = lock(r.lock) do
+            @atomicreplace r.v nothing => Some(v_local)
+        end
+        # status == true - when value obtained through call_on_owner, put! done from a different worker
+        # status == false - any other situation: atomicreplace fails, because by the time the lock is obtained cache will be populated
+        # why? put! performs caching and putting into channel under r.lock
+
+        # for local put! use the cached value, for call_on_owner cases just take the v_local as it was just cached in r.v
+        v_cache = status ? v_local : v_old
```

Comment:
I am not sure here. Since `fetch(rv.c)` succeeded, it seems like that set the value for everyone, but now we might decide to copy it locally too, but that we must have strictly maintained that `rv.v === r.v`, since `r.v` is just a locally cached copy of the same pointer from `rv.v` (which has the canonical copy) in this case. Is that right?

Reply:
Yes, that's correct. The `rv.v === r.v` is maintained right now properly I think